pkg/controllers/provisioning/scheduling/topology.go

```diff
@@ -18,149 +18,396 @@ import (
 	"context"
 	"fmt"
 	"math"
-	"strings"
 
-	"github.com/Pallinder/go-randomdata"
-	"github.com/mitchellh/hashstructure/v2"
+	"go.uber.org/multierr"
 	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/selection"
 	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
-	"github.com/aws/karpenter/pkg/utils/functional"
 	"github.com/aws/karpenter/pkg/utils/pod"
 )
 
 type Topology struct {
 	kubeClient client.Client
+	topologies map[uint64]*TopologyGroup
+	// Anti-affinity works both ways (if a zone has a pod foo with anti-affinity to a pod bar, we can't schedule bar to
+	// that zone, even though bar has no anti affinity terms on it. For this to work, we need to separately track the
+	// topologies of pods with anti-affinity terms, so we can prevent scheduling the pods they have anti-affinity to
+	// in some cases.
+	inverseTopologies map[uint64]*TopologyGroup
+	// Universe of options used to compute domains
+	requirements *v1alpha5.Requirements
 }
 
-// Inject injects topology rules into pods using supported NodeSelectors
-func (t *Topology) Inject(ctx context.Context, constraints *v1alpha5.Constraints, pods []*v1.Pod) error {
-	// Group pods by equivalent topology spread constraints
-	topologyGroups := t.getTopologyGroups(pods)
-	// Compute spread
-	for _, topologyGroup := range topologyGroups {
-		if err := t.computeCurrentTopology(ctx, constraints, topologyGroup); err != nil {
-			return fmt.Errorf("computing topology, %w", err)
+func NewTopology(kubeClient client.Client, requirements *v1alpha5.Requirements) *Topology {
+	return &Topology{
+		kubeClient:        kubeClient,
+		requirements:      requirements,
+		topologies:        map[uint64]*TopologyGroup{},
+		inverseTopologies: map[uint64]*TopologyGroup{},
+	}
+}
+
+// Initialize scans the pods provided and creates topology groups for any topologies that we need to track based off of
+// topology spreads, affinities, and anti-affinities specified in the pods.
+func (t *Topology) Initialize(ctx context.Context, pods ...*v1.Pod) (errs error) {
+	errs = multierr.Append(errs, t.updateInverseAffinities(ctx))
+	for i := range pods {
+		errs = multierr.Append(errs, t.Update(ctx, pods[i]))
+	}
+	return errs
+}
+
+func (t *Topology) Update(ctx context.Context, p *v1.Pod) (errs error) {
+	for _, topology := range t.topologies {
+		if topology.IsOwnedBy(p.UID) {
+			topology.RemoveOwner(p.UID)
```

Comment:
Do we need to handle garbage collection where the topology no longer has owners? I guess we potentially don't care since the object doesn't live longer than the batch.

Reply:
Yeah, it's short lived.  We actually can't check it here either and would need to rescan at the end of the method.  It may go to zero if the pod was the only owner and the term wasn't relaxed. In which case we would remove it, re-create it, re-query the cluster to get counts, and then add the pod back as the owner.